src/arcball-camera/ArcBall.h

```diff
@@ -0,0 +1,132 @@
+    this software dedicate any and all copyright interest in the software to
+    the public domain. We make this dedication for the benefit of the public
+    at large and to the detriment of our heirs and successors. We intend this
+    dedication to be an overt act of relinquishment in perpetuity of all
+    present and future rights to this software under copyright law.
+
+    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+    THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+    IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+    CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include <Magnum/Magnum.h>
+#include <Magnum/Math/Vector2.h>
+#include <Magnum/Math/Vector3.h>
+#include <Magnum/Math/Matrix4.h>
+#include <Magnum/Math/Quaternion.h>
+#include <Magnum/SceneGraph/MatrixTransformation3D.h>
+#include <Magnum/SceneGraph/Object.h>
+
+namespace Magnum {
+using Object3D = SceneGraph::Object<SceneGraph::MatrixTransformation3D>;
+
+/*
+ * Implementation of Ken Shoemake's arcball camera with smooth navigation feature
+ * https://www.talisman.org/~erlkonig/misc/shoemake92-arcball.pdf
+ */
+class ArcBall {
+public:
+    ArcBall(const Vector3& cameraPosition, const Vector3& viewCenter, const Vector3& upDir,
+            Deg fov, Vector2i windowSize, Object3D& cameraObject);
+
+    /* Set the camera view parameters: eye position, view center, up direction */
+    void setViewParameters(const Vector3& eye, const Vector3& viewCenter, const Vector3& upDir);
+
+    /* Reset the camera to its initial position, view center, and up dir */
+    void reset();
+
+    /* Update screen size after the window has been resized */
+    void reshape(const Vector2i& windowSize) { _windowSize = windowSize; }
+
+    /* Set the amount of lagging such that the camera will (slowly) smoothly navigate
+     * Lagging must be in [0, 1) */
+    void setLagging(Float lagging);
+
+    /* Initialize the first (screen) mouse position for camera transformation.
+     * This should be called in mouse pressed event */
+    void initTransformation(const Vector2i& mousePos);
+
+    /* Rotate the camera from the previous (screen) mouse position to the current (screen) position */
+    void rotate(const Vector2i& mousePos);
+
+    /* Translate the camera from the previous (screen) mouse position to the current (screen) mouse position */
+    void translate(const Vector2i& mousePos);
+
+    /* Translate the camera by the delta amount of (NDC) mouse position
+     * Note that NDC position must be in [-1, -1] to [1, 1] */
+    void translateDelta(const Vector2& translationNDC);
+
+    /* Zoom the camera (positive zoom values = zoom in, negative values = zoom out) */
+    void zoom(Float zoomValue);
```

Comment:
```suggestion
    void zoom(Float delta);
```

... to hint better that it's relative, not absolute